protocol-units/da-sequencer/node/src/tests/mock.rs

```diff
@@ -1,20 +1,94 @@
-	block::{BlockHeight, SequencerBlockDigest},
-	celestia::{blob::Blob, CelestiaHeight},
-	DaSequencerError, DaSequencerExternalDa, DaSequencerStorage, SequencerBlock,
-};
-use std::{cell::RefCell, future::Future};
+use crate::batch::serialize_full_node_batch;
+use crate::batch::DaBatch;
+use crate::batch::FullNodeTxs;
+use crate::block::BlockHeight;
+use crate::block::SequencerBlockDigest;
+use crate::celestia::blob::Blob;
+use crate::celestia::CelestiaHeight;
+use crate::DaSequencerError;
+use crate::DaSequencerExternalDa;
+use crate::DaSequencerStorage;
+use crate::SequencerBlock;
+use ed25519_dalek::{SigningKey, VerifyingKey};
+use futures::StreamExt;
+use movement_da_sequencer_client::{sign_batch, DaSequencerClient};
+use movement_da_sequencer_proto::blob_response::BlobType;
+use movement_da_sequencer_proto::BatchWriteRequest;
+use movement_types::block::Block;
+use movement_types::block::BlockMetadata;
+use movement_types::block::Id;
+use movement_types::transaction::Transaction;
+use std::collections::BTreeSet;
+use std::future::Future;
+use std::sync::Arc;
+use std::sync::Mutex;
+
+pub async fn mock_write_new_batch(
+	client: &mut DaSequencerClient,
+	signing_key: &SigningKey,
+	verifying_key: VerifyingKey,
+) {
+	let tx = Transaction::test_only_new(b"test data".to_vec(), 1, 123);
+	let txs = FullNodeTxs::new(vec![tx]);
+
+	//sign batch
+	let batch_bytes = bcs::to_bytes(&txs).expect("Serialization failed");
+	let signature = sign_batch(&batch_bytes, &signing_key);
+
+	// Serialize full node batch into raw bytes
+	let serialized =
+		serialize_full_node_batch(verifying_key, signature.clone(), batch_bytes.clone());
+
+	let request = BatchWriteRequest { data: serialized };
+	let res = client.batch_write(request).await.expect("Failed to send the batch.");
+	assert!(res.answer);
+}
+
+pub async fn mock_wait_and_get_next_block(
+	block_stream: &mut tonic::Streaming<movement_da_sequencer_proto::StreamReadFromHeightResponse>,
+	expected_height: u64,
+) {
+	let _ = tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
+	loop {
+		match tokio::time::timeout(std::time::Duration::from_secs(1), block_stream.next()).await {
+			Ok(Some(Ok(block))) => match block.response.unwrap().blob_type {
+				Some(BlobType::Blockv1(blockv1)) => {
+					assert_eq!(
+						blockv1.height, expected_height,
+						"stream block height 1, not the right height"
+					);
+					break;
+				}
+				Some(BlobType::Heartbeat(_)) => (),
+				None => panic!("Block stream broken at height {expected_height}"),
+			},
+			_ => panic!("No block produced at height {expected_height}"),
+		};
+	}
+}
 
 #[derive(Debug, Clone)]
-pub struct StorageMock {
-	pub batches: RefCell<Vec<DaBatch<FullNodeTxs>>>,
+pub struct StorageMockInternal {
+	pub batches: Vec<DaBatch<FullNodeTxs>>,
+	pub produced_blocks: Vec<SequencerBlock>,
 	pub current_height: u64,
+	pub parent_block_id: Id,
+}
+
+#[derive(Debug, Clone)]
+pub struct StorageMock {
+	pub inner: Arc<Mutex<StorageMockInternal>>,
```

Comment:
An `RwLock` seems more suitable here.

Reply:
In fact, I put the Mutex because it's a general object and I don't know how it's going to be used. Mutex is more general purpose, and we don't need perf here.